### PR TITLE
Mitch image and text narration integration on front end

### PR DESCRIPTION
**Pull request by mikemitchel**

### CODE REVIEW

File: client/createStory/createStory.js [hunk 1]
```diff
@@ -543,21 +569,55 @@ angular.module('storyBoard.createStory', [])
                                      playbackFinishedCallback);
   };
 
-  $scope.toggleSoundTrack = function () {
-    $scope.addSoundTrack = !$scope.addSoundTrack;
-    if ($scope.addSoundTrack) {
+  $scope.toggleSoundtrack = function () {
+    $scope.addSoundtrack = !$scope.addSoundtrack;
+    if ($scope.addSoundtrack) {
       $scope.audioTrackDisplay = "Destory Soundtrack";
```
Comment (jd009) [inline client/createStory/createStory.js]: "De-story"?  Even if it was spelled right, let's use simpler, layman terms for our users.  How about "Remove Soundtrack".

File: client/createStory/createStory.js [hunk 2]
```diff
@@ -161,6 +174,7 @@ angular.module('storyBoard.createStory', [])
     $scope.frame1ImageUrl = "http://i.imgur.com/7j15tXU.jpg";
     $scope.frame1UrlDuration = 2;
     $scope.frame1NarrationText = 'A long time ago, in a galaxy far, far away';
+    $scope.frame1NarrationDelay = 0;
```
Comment (jd009) [inline client/createStory/createStory.js]: Elsewhere, these narrationDelays are assigned as "0"; in other words, as a string type.  We should be deliberate with our types to avoid confusion, even though this is a dynamic language.

File: client/createStory/createStory.js [hunk 3]
```diff
@@ -283,7 +301,8 @@ angular.module('storyBoard.createStory', [])
           previewUrl: (wasPassed)? $stateParams.story.frames[1].previewUrl: "",
           imageUrl: $scope.frame1ImageUrl,
           imageDuration: $scope.frame1UrlDuration ? parseFloat($scope.frame1UrlDuration) : 0,
-          narrationText: $scope.frame1NarrationText
+          narrationText: $scope.frame1NarrationText,
+          narrationDelay: $scope.frame1NarrationDelay ? parseFloat($scope.frame1NarrationDelay) : 0
```
Comment (jd009) [inline client/createStory/createStory.js]: This is just a rhetorical question as a reminder that we need to think about sanitizing the inputs:  Given that these narrationDelays are strings until they are parsed as floats, what happens if the user types in words instead of numbers?  For example, a malicious user I know might type in "dog".  Again, this is an open question and a lot of our code is vulnerable to this.

File: client/createStory/createStory.js [hunk 4]
```diff
@@ -543,21 +569,55 @@ angular.module('storyBoard.createStory', [])
...
-    $scope.frame0UrlDuration = null;
   };
 
-});
+  $scope.toggleNarration = function (frameId) {
+    switch (frameId) {
+    case 1:
+      $scope.addNarration1 = !$scope.addNarration1;
+      if ($scope.addNarration1) {
+        $scope.narration1Display = "Remove Narration";
```
Comment (jd009) [inline client/createStory/createStory.js]: Could we call these $scope.narration1ButtonLabel ?

File: client/services/textToSpeechPlayer.js [hunk 5]
```diff
@@ -35,11 +38,22 @@ angular.module('storyBoard.textToSpeechPlayer', ['storyBoard.player'])
...
+          paragraphTagStr.addClass('showImagePlayerFrame');
+        }
+        window.speechSynthesis.speak(textToSpeechPlayer.utterance);
+      }, textToSpeechPlayer.narrationDelay * 1000);
+    } else {
+      if ( ! textToSpeechPlayer.isBackgroundPlayer ) {
+        var paragraphTagStr = textToSpeechPlayer.playerDiv.children();
+        paragraphTagStr.addClass('showImagePlayerFrame');
+      }
+      window.speechSynthesis.speak(textToSpeechPlayer.utterance);
```
Comment (jd009) [inline client/services/textToSpeechPlayer.js]: This is a lot of duplicated code, which makes it hard for the reader to identify the difference between the IF case and the ELSE case.  Also, it paves the way for future bugs where we update one copy but not the other.  Even if we remember to update both copies, we have to add twice as much code than if we only had one copy of this logic.
Let's wrap it up into a function (Notice the **underscore**, it is a private function that we internally use):
```
TextToSpeechPlayer.prototype._play = function() {
  if ( ! this.isBackgroundPlayer ) {
    var paragraphTagStr = this.playerDiv.children();
    paragraphTagStr.addClass('showImagePlayerFrame');
  }

  window.speechSynthesis.speak(this.utterance);
}
```
Then call the private function from our public play function:
```
TextToSpeechPlayer.prototype.play = function(){
  if(this.narrationDelay) {
    var narrationDelayInSeconds = this.narrationDelay * 1000;
    var boundPlayFunction = this._play.bind(this);
    setTimeout(
     boundPlayFunction,
     narrationDelayInSeconds
    );
  } else {
    this._play();
  }
}
```
Now, it's much more obvious what the difference is between the IF case and the ELSE case, plus we are only doing keyword **this** trickery when we need to (i.e. the IF case).

File: server/src/schema.go [hunk 6]
```diff
@@ -40,17 +40,22 @@ type Story struct {
...
+	Volume         int      `json:"volume"`
+	PreviewUrl     string   `json:"previewUrl" bson:"previewurl"`
+	ImageUrl       string   `json:"imageUrl" bson:"imageurl"`
+	ImageDuration  int      `json:"imageDuration"`
+	NarrationText  string   `json:"narrationText"`
+	NarrationDelay float32  `json:"narrationDelay"`
+	AudioId        string   `json:"audioId"`
+	AudioStart     float32  `json:"audioStart"`
+	AudioEnd       float32  `json:"audioEnd"`
+	AudioDelay     float32  `json:"audioDelay"`
```
Comment (jd009) [inline server/src/schema.go]: How did these AudioId, AudioStart, AudioEnd, and AudioDelay variables sneak into this pull request?  :-)  These variables are unrelated to the image narration feature.  Moreover, I don't think we, as a group, gave the green light for per-frame audio.  Either way, pull requests should have a single, clear purpose with no extraneous side plots (to use a story analogy).

Comment (mikemitchel) [inline server/src/schema.go]: i shared my branch so danny g could test his back end stuff, and he edited or merged his stuff into my branch

File: client/services/textToSpeechPlayer.js (hunk 5 above)
Comment (mikemitchel) [inline client/services/textToSpeechPlayer.js]: this is danny g, but will fix (See below comment)

File: client/createStory/createStory.js (hunk 1 above)
Comment (mikemitchel) [inline client/createStory/createStory.js]: that's a milco add, but I will change it, as it might have been me who said to be explicit about destroying the soundtrack.

File: client/createStory/createStory.js (hunk 3 above)
Comment (mikemitchel) [inline client/createStory/createStory.js]: good point, and would be an easier refactor to fix types.

File: client/createStory/createStory.js (hunk 4 above)
Comment (mikemitchel) [inline client/createStory/createStory.js]: we can and will

File: client/createStory/createStory.js [hunk 7]
```diff
@@ -545,21 +583,62 @@ angular.module('storyBoard.createStory', [])
...
+  $scope.toggleSoundtrack = function () {
+    $scope.addSoundtrack = !$scope.addSoundtrack;
+    if ($scope.addSoundtrack) {
+      $scope.audioButtonLabel = "Remove Soundtrack";
+=======
   $scope.toggleSoundTrack = function () {
     $scope.addSoundTrack = !$scope.addSoundTrack;
     if ($scope.addSoundTrack) {
       $scope.audioTrackDisplay = "Remove Soundtrack";
+>>>>>>> master
```
Comment (jd009) [inline client/createStory/createStory.js]: Merging artifacts.  Does this code run like this?  Or does it throw exceptions?  Be sure to do a run-through (some testing) before submitting a pull request.

File: client/createStory/createStory.js [hunk 8]
```diff
@@ -545,21 +583,62 @@ angular.module('storyBoard.createStory', [])
                                      playbackFinishedCallback);
   };
 
+<<<<<<< HEAD
```
Comment (jd009) [inline client/createStory/createStory.js]: <<<<<<<< HEAD, merge conflict artifact

File: client/services/textToSpeechPlayer.js [hunk 9]
```diff
@@ -42,6 +53,7 @@ angular.module('storyBoard.textToSpeechPlayer', ['storyBoard.player'])
     window.speechSynthesis.speak(this.utterance);
   };
 
+
```
Comment (jd009) [inline client/services/textToSpeechPlayer.js]: Extra line of spacing.

File: client/createStory/createStory.js [hunk 10]
```diff
@@ -371,7 +407,8 @@ angular.module('storyBoard.createStory', [])
           volume: $scope.frame2Volume,
           imageUrl: $scope.frame2ImageUrl,
           imageDuration: $scope.frame2UrlDuration,
-          narrationText: $scope.frame2NarrationText
+          narrationText: $scope.frame2NarrationText,
+          narrationDelay: $scope.frame1NarrationDelay
```
Comment (jd009) [inline client/createStory/createStory.js]: This is frame 1's narration delay being assigned to frame 2.
